jans-cedarling/cedarling/src/jwt/mod.rs

```diff
@@ -90,37 +95,72 @@ impl JwtService {
         access_token: &str,
         id_token: &str,
         userinfo_token: &str,
-    ) -> Result<(A, I, U), Error>
+    ) -> Result<(A, I, U), JwtDecodingError>
     where
         A: DeserializeOwned,
```

Comment:
I think `DeserializeOwned` as a return type here is too general. Is there perhaps a type in our code which adequately captures token claims?

I also think that extracting the claims before decoding (and optionally validating) the tokens is a little odd. At best there's some redundancy since the purpose of decoding is to extract claims anyway. At worst it's an invitation to logic errors in relation to token decoding.

And since `extract_claims` also returns a `DeserializeOwned`, I suspect the problem is related to that. I'm also thinking that `extract_claims` should be operating on a `token::Whatever` rather than `&str`. OTOH having something like `extract_claims` acting on a `&str` is perhaps a good candidate for a convenience function.

However, I realise we're aiming at MVP soon, so I don't want those concerns to be blockers. But I do think they need to be addressed.

Reply:
i was planning to fix these things in #9967. i thought i'd make the PRs as small as possible only focusing on certain problems at a time so it would be easier to review... but yeah, what happens usually is people see things they don't like at the current state without really any context on the future plans 😅

Let me know if you think it's more beneficial to just bunch all of the changes in one PR

Reply:
> I think DeserializeOwned as a return type here is too general. Is there perhaps a type in our code which adequately captures token claims?
The thing is, i was told that the user would be defining the claims inside the Tokens. I think it might also change based on some policies defined in `policy_store.json`.

> I also think that extracting the claims before decoding (and optionally validating) the tokens is a little odd. At best there's some redundancy since the purpose of decoding is to extract claims anyway. At worst it's an invitation to logic errors in relation to token decoding.
Yes, i agree, the issue is probably we need to perform some checks which we assume that any token should contain plus the custom payloads.... while i was writing this down, i've thought of a solution so i updated #9967

Reply:
Sounds good to me.